### PR TITLE
Put Give Onchain

### PR DESCRIPTION
**Pull request by crabsinger**

### CODE REVIEW

File: _api/hasura/cron/giveOnchainSyncer.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
+/* eslint-disable no-console */
```
Comment (CryptoGraffe) [inline _api/hasura/cron/giveOnchainSyncer.ts:1]: hmmm

File: api-lib/gql/__generated__/zeus/const.ts [hunk 2]
```diff
@@ -2204,15 +2204,18 @@ export const AllTypesProps: Record<string, any> = {
     _not: 'colinks_gives_bool_exp',
     _or: 'colinks_gives_bool_exp',
     activity_id: 'bigint_comparison_exp',
+    attestation_uid: 'String_comparison_exp',
     cast_hash: 'String_comparison_exp',
     created_at: 'timestamptz_comparison_exp',
     give_skill: 'skills_bool_exp',
     giver_profile_public: 'profiles_public_bool_exp',
     id: 'Int_comparison_exp',
+    onchain_sync_error: 'String_comparison_exp',
```
Comment (CryptoGraffe) [inline api-lib/gql/__generated__/zeus/const.ts:2213]: maybe we want on onchain_synced_at timestamp

File: _api/hasura/cron/giveOnchainSyncer.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+async function giveToSync() {
+  // get all give that is not on-chain
+  const { colinks_gives } = await adminClient.query(
+    {
+      colinks_gives: [
+        {
+          where: {
+            tx_hash: { _is_null: true },
```
Comment (CryptoGraffe) [inline _api/hasura/cron/giveOnchainSyncer.ts:117]: what about ones with errors? could we get stuck forever?

File: _api/hasura/cron/giveOnchainSyncer.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+          {
+            operationName: 'cron_giveOnchainSyncer__attestGive',
+          }
+        );
+
+        success++;
+      } catch (e: any) {
+        errors++;
+
+        await adminClient.mutate(
```
Comment (CryptoGraffe) [inline _api/hasura/cron/giveOnchainSyncer.ts:65]: should this just happen in the attest function so the updates are all in one place

File: api-lib/eas.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  '0x44ba9714fe382b86f0e126a592116bc2fc45905ee034d0e8b5db218bf366dafe';
+const BASE_SEPOLIA_SCHEMA_UID =
+  '0x55df90070189848b41fc4f2a8e07f82d6434718696286b9188ceb1d737f6cf87';
+
+export async function attestGiveOnchain(give: Give) {
+  const chainId = Number(baseChain.chainId);
+  const provider = getProvider(chainId);
+
+  const syncerWallet = new Wallet(COSOUL_SIGNER_ADDR_PK);
+  const signer = syncerWallet.connect(provider);
```
Comment (CryptoGraffe) [inline api-lib/eas.ts:25]: do we want a diff addr for this or is this all good

File: api-lib/eas.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  );
+  const encodedData = schemaEncoder.encodeData([
+    {
+      name: 'from',
+      value: giverAddr,
+      type: 'address',
+    },
+    { name: 'skill', value: give.skill, type: 'string' },
+    { name: 'amount', value: 1, type: 'uint16' },
+    { name: 'context', value: context, type: 'string' },
```
Comment (CryptoGraffe) [inline api-lib/eas.ts:53]: context still feels ambiguous

File: api-lib/gql/__generated__/zeus/const.ts (hunk 2 above)
Comment (crabsinger) [reply api-lib/gql/__generated__/zeus/const.ts:2213]: ok, will add and use as primary search for things to sync

File: _api/hasura/cron/giveOnchainSyncer.ts (hunk 3 above)
Comment (crabsinger) [reply _api/hasura/cron/giveOnchainSyncer.ts:117]: yeah... we could, any good ways around this?

File: api-lib/eas.ts (hunk 5 above)
Comment (crabsinger) [reply api-lib/eas.ts:26]: i think just this one and then we have onchain.coordinape.eth set to it

File: hasura/migrations/default/1717008407293_alter_table_public_colinks_gives_add_column_tx_hash/up.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+alter table "public"."colinks_gives" add column "tx_hash" text
```
Comment (sshmm) [inline hasura/migrations/default/1717008407293_alter_table_public_colinks_gives_add_column_tx_hash/up.sql:1]: is not it better to add all changes of the same table of the same feature to the same migration file?

File: _api/hasura/cron/giveOnchainSyncer.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    // eslint-disable-next-line no-console
+    console.log('Syncing complete', { success, errors });
+    res.status(200).json({
+      success: true,
+      gives: gives.length,
+      errorsCount: errors,
+      successCount: success,
+    });
+  } catch (e: any) {
```
Comment (sshmm) [inline _api/hasura/cron/giveOnchainSyncer.ts:33]: will this "try ..catch" catch any errors? don't we need to throw again if errors is greater than 0

Comment (crabsinger) [reply _api/hasura/cron/giveOnchainSyncer.ts:33]: good catch - moved the give fetching into the block

File: hasura/migrations/default/1717008407293_alter_table_public_colinks_gives_add_column_tx_hash/up.sql (hunk 7 above)
Comment (crabsinger) [reply hasura/migrations/default/1717008407293_alter_table_public_colinks_gives_add_column_tx_hash/up.sql:1]: it is better - but im lazy now

**Pull request merged**